text/0000-PXE-Booting-UEFI.md

```diff
@@ -0,0 +1,228 @@
+PXEGRUB |	PXEGrub |	UEFI |	grub/bootia32.efi, grub/bootx64.efi |	grub/01-AA-BB-CC-DD-EE-FF |	grub/efidefault |	New
+PXEGRUB2 |	PXEGrub |	BIOS | pxegrub2 |	grub2/grub.cfg-aa-bb-cc-dd-ee-ff | grub2/grub.cfg |	PXELinux will be used instead usually, but default filename will be set for generic OS.
+PXEGRUB2 |	PXEGrub |	UEFI |	grub2/grubx32.efi, grub2/grubx64.efi |	grub2/grub.cfg-aa-bb-cc-dd-ee-ff |	grub2/grub.cfg |	New
+POAP | POAP | - |	poap.cfg/aabbccddeeff |	poap.cfg/AABBCCDDEEFF |	- |	No changes
+ZTP |	ZTP |	- |	poap.cfg/aabbccddeeff |	poap.cfg/AABBCCDDEEFF |	- |	No changes
+
+For UEFI systems, 32bit or 64bit EFI loader will be configured on DHCP
+according host's architecture association. Most Linux distributions have
+limited or zero support of 32bit EFI (loaders are simply not shipped in their
+repositories), but to give the possibility of booting native 32bit EFI loaders
+filenames "bootia32.efi" and "grubx32.efi" will be handed over by DHCP in order
+to support that path.
+
+In secure boot environments, loaders must be either signed or an alternative
+shim loader can be used. Users will be able to select third flag called "Secure
+Boot UEFI" which will cause setting DHCP filename to either "shim.efi",
+"grub/shim.efi" or "grub2/shim.efi" depending on variant which will then pass
+the check and chain-load unsigned firmware named "bootx64.efi" or "grubx64.efi"
+from the same directory.
+
+PXEGrub (version 1) configuration path is changing from boot/grub/ to grub/ for
+consistency, a change in Foreman core will be needed but no migration or manual
+change after upgrade is needed - existing hosts in boot mode will continue
+booting from the old dir, new ones will use the new path. TFTP file deletion
+does not trigger error, thus the only drawback is there might be configuration
+files left after some time. We will put this into upgrade instructions.  Edit
+
+## Implementation
+
+Both Foreman Core and Smart Proxy changes are required in order to support all
+of the above.
+
+### [Foreman proxy #12633](http://projects.theforeman.org/issues/12633) - Add Pxegrub2 variant and allow multiple TFTP config files
+
+Foreman Proxy must be enhanced with new PXE variant "Pxegrub2" to allow
+deploying configuration files in "grub2/" folder. In addition to that, Pxegrub
+(version 1) must deploy two configuration files instead of one now to support
+both BIOS and UEFI naming conventions (grub/menu.lst vs grub/efidefault and
+grub/menu.lst.01aabbccddeeff vs grub/01-AA-BB-CC-DD-EE-FF). Also, Pxegrub
+variant will use new base directory "grub/" instead of "boot/grub/" for
+consistency with "grub2" and "pxelinux.cfg/".
+
+### [Foreman core #12634](http://projects.theforeman.org/issues/12634) - New host/hostgroup flags pxe_template and loader
+
+Host and hostgroups must be enhanced with two new flags: PXE template specifies
+PXE template kind and variant pair to use. Until today, this was hardcoded to
```

Comment:
> my point is that it seems like its going to be settings that are  not unique per host, rather per a group of hosts, so its either the hw model, or a new join table (e.g. os profile) that should accept it as a paramter.
> I know its already overloaded, hence I'm hoping an OS profile (or maybe some other object) is a better name to represent that instead of keep on enlarging host/groups. after all, we are working hard to reduce the size of the host object using facets with @shimshtein

I can introduce provision profile with the following fields:

* architecture
* OS
* ptable
* media

The thing is I would like to hit the 1.12 timeframe with this. This is a
huge refactoring. In this light, attaching it to the hardware model
seems like a better idea, we can add similar global setting to avoid
puppet fact to override the setting.

Others have opinions?

-- 
Later,
 Lukas #lzap Zapletal